embedded_rtic/examples/sonars.rs

```diff
@@ -0,0 +1,440 @@
+
+        let times = [Instant::from_ticks(0);3];
+
+        let token = rtic_monotonics::create_nrf_timer0_monotonic_token!();
+        Mono::start(cx.device.TIMER0, token);
+
+        let _ = trigger_trampoline1::spawn();
+        let _ = trigger_trampoline2::spawn();
+        let _ = trigger_trampoline3::spawn();
+
+        (Shared { gpiote }, Local {
+            trig,
+            echo,
+            trig2,
+            echo2,
+            trig3,
+            echo3,
+            sender,
+            receiver,
+            sender2,
+            receiver2,
+            sender3,
+            receiver3,
+            previous_time,
+            previous_time2,
+            previous_time3,
+            times,
+        })
+    }
+
+    const NUM_SONARS: usize = 3; // Number of sonars
+    #[derive(Copy, Clone)]
+    enum SonarChannel {
+        Forward,
+        Left,
+        Right,
+    }
+
+    #[task(binds = GPIOTE,priority = 4, local=[echo,sender,previous_time,echo2,sender2,previous_time2,echo3,sender3,previous_time3,times],shared = [gpiote])]
+    /// Reads the echo pin
+    ///
+    /// When ever the echo pin goes high or low this interrupt is triggered.
+    /// On the falling edge this interrupt computes the "width" in time
+    /// of the square wave thus allowing us to compute the time it took
+    /// to echo back to us.
+    fn echo(mut cx: echo::Context) {
+        // Get time ASAP for highest granularity
+        let time = Mono::now();
+        let zero: Instant = Instant::from_ticks(0);
+
+        // Array to store triggered sonars
+        let mut triggered_sonars = [None; NUM_SONARS];
+
+        // Check which sonar triggered the event and store it in the array
+        let debug_time = time.duration_since_epoch().to_micros();
+        cx.shared.gpiote.lock(|gpiote| {
+            if gpiote.channel0().is_event_triggered() {
+                info!("Interrupt on channel 0, {:?}", debug_time);
+                triggered_sonars[0] = Some(SonarChannel::Forward);
+            }
+            if gpiote.channel1().is_event_triggered() {
+                info!("Interrupt on channel 1, {:?}", debug_time);
+                triggered_sonars[1] = Some(SonarChannel::Left);
+            }
+            if gpiote.channel2().is_event_triggered() {
+                info!("Interrupt on channel 2, {:?}",debug_time);
+                triggered_sonars[2] = Some(SonarChannel::Right);
+            }
+        });
+
+        // Print messages for each triggered sonar
+        for sonar in triggered_sonars.iter().filter_map(|x| *x) {
+            match sonar {
+                SonarChannel::Forward => {
+                    //info!("Event triggered by Sonar 1 (Forward)");
+                    if cx.local.times[0] == zero {
+                        cx.local.times[0] = time;
+                    } else {
+                        let distance = time.checked_duration_since(cx.local.times[0]).unwrap().to_micros() / 56;
```

Comment:
might be worth breaking this out in to a separate function
```rust

    pub fn compute_and_send_distance<'a>(time:u32, sender:&mut Sender<'a,u32,CAPACITY>){
        let distance = time / 56;
        match sender.try_send(distance as u32) {
            Ok(_) => {}
            _ => {
                error!("Distance did not fit in to the buffer, scheduling is probably broken.");
                // This is not fatal so we simply continue with our life
            }
        }
    }
    ```
    To ensure that all of the tree implementations follow the same behaviour
    
    This could/should be placed in file
    ```bash
    | - src
       | - lib.rs
       | - helpers.rs # New file
```

It might even be better to re-write that function into a more testable format as this implementation both sends data and computes things it might be better to do this in some other way, I just think that it is worth breaking it out since we will use the same approach in the real app.